File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    public String[] solution(String[][] tickets) {
        List<String> result = new ArrayList<>();
        Map<String, PriorityQueue<String>> map = new HashMap<>();
        
        for (String[] ticket : tickets) {
            map.computeIfAbsent(ticket[0], k -> new PriorityQueue<>()).add(ticket[1]);
        }
        
        dfs(result, map, "ICN");
        Collections.reverse(result);
        
        return result.toArray(String[]::new);
    }
    
    private void dfs(List<String> result, Map<String, PriorityQueue<String>> map, String from) {
        while (map.containsKey(from) && !map.get(from).isEmpty()) {
            dfs(result, map, map.get(from).poll());
        }
        
        result.add(from);
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.kt
import java.util.*;

class Solution {
    fun solution(tickets: Array<Array<String>>): Array<String> {
        val result = mutableListOf<String>()
        val map = mutableMapOf<String, PriorityQueue<String>>()
        
        for (ticket in tickets) {
            map.getOrPut(ticket[0]) { PriorityQueue() }.add(ticket[1])
        }
        
        fun dfs(from: String) {
            while (map.containsKey(from) && map[from]!!.isNotEmpty()) {
                dfs(map[from]!!.poll())
            }
            
            result.add(from)
        }
        
        dfs("ICN")
        return result.reversed().toTypedArray()
    }
}